services-js/registry-certs/client/models/UploadableFile.ts

```diff
@@ -0,0 +1,190 @@
+  | 'canceling'
+  | 'deleting'
+  | 'deleted'
+  | 'uploadError'
+  | 'deletionError';
+
+// todo: ie11 not displaying files in supporting docs list
+
+/**
+ * Binary files provided by the user are immediately uploaded to (or deleted
+ * from) the db server. Because an input may accept multiple files at once,
+ * each UploadableFile is responsible for uploading or deleting its File on the
+ * server.
+ */
+export default class UploadableFile {
+  private readonly fetchGraphql: FetchGraphql;
+  readonly file: File;
+  private readonly uploadSessionId: string;
+
+  @observable status: Status;
+  @observable progress: number; // 0 - 100
+
+  @observable.ref uploadRequest: XMLHttpRequest | null = null;
+  @observable.ref uploadResponse: UploadResponse | null = null;
+  @observable errorMessage: string | null = null;
+
+  constructor(
+    file: File,
+    uploadSessionId: string,
+    status?: Status,
+    progress?: number
+  ) {
+    this.fetchGraphql = fetchGraphql();
+
+    this.file = file;
+    this.uploadSessionId = uploadSessionId;
+    this.status = status || 'idle';
+    this.progress = progress || 0;
+  }
+
+  @action
+  upload() {
+    const uploadRequest = new XMLHttpRequest();
+    const formData = new FormData();
+
+    formData.append('file', this.file);
+    formData.append('type', 'BC');
+    formData.append('uploadSessionId', this.uploadSessionId);
+
+    uploadRequest.onload = this.handleLoad;
+    uploadRequest.onerror = this.handleError;
+    uploadRequest.onloadend = this.handleLoadEnd;
+
+    if (uploadRequest.upload) {
+      uploadRequest.upload.onprogress = this.handleProgress;
+    }
+
+    this.progress = 0;
+    this.status = 'uploading';
+
+    uploadRequest.open('POST', '/upload');
+    uploadRequest.send(formData);
+  }
+
+  // If the user is trying to cancel an in-progress upload, we want to express
+  // the appropriate status while the cancellation is pending.
+  @action
+  async delete(didCancel?: boolean): Promise<boolean> {
+    if (!this.attachmentKey) {
+      return false;
+    }
+
+    didCancel ? (this.status = 'canceling') : (this.status = 'deleting');
+
+    try {
+      const result = await deleteBirthCertificateUploadedFile(
+        this.fetchGraphql,
+        this.uploadSessionId,
+        this.attachmentKey
+      );
+
+      if (result.success) {
+        runInAction(() => {
+          this.status = 'deleted';
+        });
+
+        return true;
+      } else {
+        throw new Error(result.message || 'Unknown error');
+      }
+    } catch (e) {
+      runInAction(() => {
+        this.status = 'deletionError';
+      });
+
+      return false;
+    }
+  }
+
+  @action.bound
+  handleLoad(ev: ProgressEvent) {
+    const xhr = ev.target;
+    let json: any = null;
+
+    if (!(xhr instanceof XMLHttpRequest)) {
+      return;
+    }
+
+    try {
+      json = JSON.parse(xhr.responseText);
+    } catch (e) {
+      this.status = 'uploadError';
+      this.errorMessage = xhr.statusText;
```

Comment:
should probably `return` here since JSON not parsing is an error in and of itself.